x/staking/cache/cache.go

```diff
@@ -0,0 +1,188 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+type slice[T any] interface {
+	~[]T
+}
+
+type cacheEntry[K comparable, V slice[T], T any] struct {
+	mu   sync.RWMutex
+	data map[K]V
+	// cache would be invalidated if inserted data is greater than max
+	invalidated bool
+	// max defines the maximum number of entries in each cache map
+	// to prevent OOM attacks.
+	// - if max == 0, there is no cap on the number of entries in the cache
+	// - if max > 0, the cache will cap the number of entries it stores
+	// - if max < 0, the cache is a no-op cache.
+	max int
+}
+
+func newCacheEntry[K comparable, V slice[T], T any](max int) *cacheEntry[K, V, T] {
+	return &cacheEntry[K, V, T]{max: max}
+}
+
+func (e *cacheEntry[K, V, T]) get() (map[K]V, bool) {
+	if e.max < 0 {
+		return nil, false
+	}
+
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	if e.data == nil {
+		return nil, e.invalidated
+	}
+
+	copied := make(map[K]V, len(e.data))
+	for k, v := range e.data {
+		copied[k] = append([]T(nil), v...)
+	}
+
+	return copied, e.invalidated
+}
+
+func (e *cacheEntry[K, V, T]) set(data map[K]V) {
+	if e.max < 0 {
+		return
+	}
+
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	totalElements := 0
+	for _, v := range data {
+		totalElements += len(v)
+	}
+
+	if e.max > 0 && totalElements > e.max {
+		e.invalidated = true
+		return
+	}
+
+	copied := make(map[K]V, len(data))
+	for k, v := range data {
+		if len(v) == 0 {
+			continue
+		}
+		sliceCopy := append([]T(nil), v...)
+		copied[k] = sliceCopy
+	}
+
+	e.data = copied
+	e.invalidated = false
+}
+
+func (e *cacheEntry[K, V, T]) setEntry(key K, value V) {
+	if e.max < 0 {
+		return
+	}
+
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
```

Comment:
add early return check
```suggestion
if e.invalidated {
     return
}
```